[IMP] web_responsive: Add search feature

yajo opened the pull request:
Indludes a squashed and slightly altered version of @lasley's #838, and supersedes that stale PR.

You should be able to search menus now.

@Tecnativa

lasley committed:
[IMP] web_responsive: Add search feature (#838)

* Add feature to search menus in the app drawer
* Switch model to RPC

yajo committed:
[FIX] web_responsive: Complete search feature forward-porting

pedrobaeza commented on web_responsive/readme/DESCRIPTION.rst:
```diff
@@ -2,7 +2,22 @@ This module provides a mobile compliant interface for Odoo Community web.
+* Set chatter side (Optional per user)
+* Quick search (see below)
+
+The following keyboard shortcuts are implemented:
```
This should be in "Usage" section, isn't it?

pedrobaeza reviewed (dismissed):
Travis is red, and #850 is also present here. There should a method that returns the available menus instead of searching the ir.ui.menu entries.

yajo committed:
[FIX] web_responsive: Disable access to hidden menus

When doing menu search, some menus that the user didn't have permission to use were presented to him.

Fix #850.

pedrobaeza commented:
Please solve the conflict now that I have merged the other PR, and can you try to find a cleanest way of solving the menu issue. Standard menu and enterprise doesn't need these methods.

pedrobaeza commented:
@Yajo what is the status of this?

yajo commented:
I rebased, I wasn't aware of the git conflicts.

Current implementation is worse than enterprise because, as you said, it should use the cached menus instead of asking the server, but it's much better than it was in v10 because it defers the search until user stops writing and it aborts current search if he starts writing again, and because it fixes #850.

Implementing enterprise-like search would probably mean refactoring the addon, which is actually quite messy.

All of this is written in the known issues section. I wouldn't personally block enhancements just for not being perfect, mostly when they are much better than what was there before, but I'm not the one to decide on that. If you prefer to block it until it's perfect, just be aware it will take some time.

Besides all of that, it's still missing reviews.

pedrobaeza commented:
Using this PR, half of the menus disappear not in the search, but on the top part. Can you please check?

yajo commented:
I cannot reproduce :confused:

pedrobaeza commented:
Put it in our instance as a test and you'll see the problem.

Tardo commented:
Some error on runbot:
![error_odoo](https://user-images.githubusercontent.com/731270/45303586-70223c80-b516-11e8-8bb6-d6f085d682ca.png)

Login as "demo" user cause mistakes...
- When no app available in the menu, if use search feature and press enter you get an "undefined" url.
- When app available you can't navigate... do a click on icon doesn't launch any action.

yajo commented:
> Some error on runbot:

This comes from the `web_action_conditionable` addon, which asking for a dependency that doesn't exist (`web.View`). It's unrelated AFAICS.



> When app available you can't navigate... do a click on icon doesn't launch any action.

It seems to be a cache problem. Reloading without cache worked for me. Can you check please?

> When no app available in the menu, if use search feature and press enter you get an "undefined" url.

It should be fixed now.

yajo commented:
> half of the menus disappear not in the search, but on the top part

I finally managed to reproduce this, I'll work on the fix. Thanks!

yajo commented:
It should work now.

tarteo reviewed (changes requested):
Cool feature!

Is it possible to search on the display_name instead of the name e.g. 'Settings' should also find menu 'Settings/Dashboard'.

yajo commented:
It's not possible because it is a computed, non-stored, context-dependant field.

The only way would be doing an offline search, purely in JS, through the cached visible menus, as `web_enterprise` does, but that'd mean a refactor of this addon, which is needed anyway BTW, but honestly I'd try to tackle that when migrating to v12.

pedrobaeza reviewed (approved):
Please squash commits before merging

pedrobaeza commented:
I have squashed already, so I merge.
